serializers/src/main/java/io/pravega/schemaregistry/serializers/BaseDeserializer.java

```diff
@@ -0,0 +1,21 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.serializers;
+
+import io.pravega.client.stream.Serializer;
+
+import java.nio.ByteBuffer;
+
+abstract class BaseDeserializer<T> implements Serializer<T> {
+    @Override
+    public final ByteBuffer serialize(T value) {
```

Comment:
I understand that the writer serializes and the reader deserializes in Pravega, but the pattern we have been using is that we pass a `Serializer` implementation that contains both. Why do we need to create this separation with the registry?

Reply:
Serializers require schema. 
For several deserializers (esp generic deserializers) the schema is optional. 

A serializer config could be different from deserializer.
For example:
Serializers require Encoder.
Desializers require Decoders.
There is a flag - fail on codec mismatch which is relevant for deserializers. 

It is cleaner to explicitly state whether user wants to create a serializer or deserializer and initialize it with specific config.

Reply:
I'm confused about this separation of encoders and decoders, in the end, we should be calling them codecs. I'm not sure I follow why failing on codec mismatch is only relevant while reading. Does it mean that when we write, we accept any codec that is given independent of the current encoding being used?

Reply:
> I'm not sure I follow why failing on codec mismatch is only relevant while reading. 

A writer uses a single encoder. So it can use a codec which is registered. 
A reader can receive events written by multiple writers. Thereby they could be encoded with different codecs. And hence it needs different decoders. 
So we provide a flag to indicate if the application wants to prevent a start of a reader if it doesnt have decoders for all codecs that are registered with the service.

Reply:
The reason to keep them separate while creating a "codec" which extends from the two is as follows:
Before encoding the data, a CodecType is registered with the service. 
This has two things -
1. name
2. a map of properties. 

During deserializers, users can supply their decoder functions for each codecType "name" registered with the service. 

When deserializer sees the data with encoding id, it gets the corresponding encodingInfo object which has schema and codecType. 
This includes name and the properties map. 

The decode method corresponding to the codecType "name" is called with encoded bytes and map of strings (properties map).
Additional information about the encoding could have been shared using the properties map. 
The decoder can make use of properties map optionally for decoding.